### ISSUE TITLE
Repo: CI runs on `main` commits are being cancelled if a new commit lands before CI completes

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#8773]
### Suggestion

https://github.com/typescript-eslint/typescript-eslint/pull/7819

This PR added concurrency config for our actions. This is great for branch (PR) CI runs as it ensures we don't keep CI runs on old commits if a new commit comes in. So if a user spans a bunch of pushes then we'll only have one run per PR.

However the config also applies to `main` - which is a problem. We rely on the `main` CI to validate each commit is correct. We also rely on it to publish a canary version per commit.

This has lead us into this state:

![Screenshot of man's history](https://github.com/typescript-eslint/typescript-eslint/assets/7462525/b5b256be-4899-45e7-b6ef-9808ab79fdf4)

We currently have a breakage in the website caused by a semantic merge conflict.
However we have no idea what commit caused the problem because so many commits have had their CI cancelled.

We need to exclude `main` from this concurrency config.

**Comment by JoshuaKGoldberg** [comment]
Related / byproduct: #8774 -> #8775.

**Comment by JoshuaKGoldberg** [comment]
Hmm, https://docs.github.com/en/actions/using-jobs/using-concurrency isn't promising.

Maybe we should enable the dreaded _require branches be up to date_ option?

**Comment by bradzacher** [comment]
https://docs.github.com/en/actions/using-jobs/using-concurrency#example-only-cancel-in-progress-jobs-on-specific-branches

Looks like you can use conditions in the flag so we can disable the cancelling just on the `main` branch.

```yaml
concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: ${{ !contains(github.ref, 'release/')}}
```
